[13.0][IMP] project_timesheet_time_control: add a field for the end time of a timesheet

davejames opened the pull request:
This commit adds a field to the timesheet to store the end time. It is not required to enter the end time, if it is not entered it is calculated automatically by comparing the start time and the duration, if only a start time and end time are entered, then the duration is calculated accordingly. The end time field is set to optional in the tree view.

astirpe commented on project_timesheet_time_control/views/account_analytic_line_view.xml:
```diff
@@ -94,7 +96,8 @@
             </xpath>
             <field name="date" position="after">
-                <field name="date_time" required="1" />
+                <field name="date_time" />
```
Why removing the `required="1"` here?

davejames replied:
this is a kanban view, so required is redundant

astirpe commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -13,13 +15,33 @@ class AccountAnalyticLine(models.Model):
+    def _compute_date_time_end(self):
+        for record in self:
+            record.date_time_end = record.date_time + relativedelta(
+                hours=record.unit_amount
```
I have the feeling that the UOM should be included in the compute of `unit_amount`, to avoid errors case of UOMs other than time.

astirpe commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -13,13 +15,33 @@ class AccountAnalyticLine(models.Model):
+
+    def _inverse_date_time_end(self):
+        for record in self:
+            record.unit_amount = (
```
Here you are assuming that `unit_amount` is always in UOM expressed in hours, that could be not always the case.

davejames replied:
yeah you're right, i hadn't thought about non-timesheet related usage of this model

pedrobaeza replied:
There's no sense in filling these fields for non-timesheet analytic lines. What has to be taken into account (it's also a mistake in previous module state), is if the timesheet unit is not hours, but days (or other compatible UoM).

davejames replied:
I just pushed a commit to check if the uom is hourly, i think if the uom is daily then the end time isn't necessary anyway

pedrobaeza replied:
I think it is, as it should put the fraction of the working hour day. Please check how core does when you select days as timesheet unit.

davejames replied:
I don't see a lot of value in calculating the days. For example, you may put a quantity of 5 days, which would be converted into 40 hours by default, so automatically setting the end date would need to take the working schedule into account, but in reality it's guess work, that 5 days might not be a contiguous period of time.

pedrobaeza reviewed (approved):
We can deprecate right now the other module with this.

astirpe commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -13,13 +15,39 @@ class AccountAnalyticLine(models.Model):
+                    hours=record.unit_amount
+                )
+            else:
+                record.date_time_end = False
```
I think it should be like this:

```suggestion
                record.date_time_end = record.date_time_end
```

davejames replied:
included in my latest commit

astirpe commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -13,13 +15,39 @@ class AccountAnalyticLine(models.Model):
+    def _compute_date_time_end(self):
+        hour_uom = self.env.ref("uom.product_uom_hour")
+        for record in self:
+            if record.product_uom_id == hour_uom:
```
You should also check that both `date_time` and `unit_amount` have a value to calculate `date_time_end`

davejames replied:
included in my latest commit

astirpe commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -13,13 +15,43 @@ class AccountAnalyticLine(models.Model):
         help="Indicate which time control button to show, if any.",
     )
 
+    @api.depends("date_time", "unit_amount")
```
UOM missing in the depends

davejames replied:
uom is already a dependency of hr_timesheet

astirpe replied:
I mean the dependency of the compute method

davejames replied:
oooh, yes you're right

davejames replied:
included in latest commit

astirpe commented on project_timesheet_time_control/views/account_analytic_line_view.xml:
```diff
@@ -104,6 +107,9 @@
             </xpath>
             <xpath expr="//templates//t[@t-esc='record.date.value']" position="after">
                 <t t-esc="record.date_time.value" />
+                <br />
```
Not sure if this `<br />` is intentional. I think it shouldn't be here.

astirpe replied:
Shouldn't this block:

```
               <span>to</span>
                <t t-esc="record.date_time_end.value" />
```

be displayed only if `date_time_end` has a value?

davejames replied:
it's intentional, it's so that the times are both in line with each other. This is easier on the eyes in my opinion than having them on the same line.

![image](https://user-images.githubusercontent.com/1944663/90889619-d5b61d80-e3fb-11ea-9964-87851b27ef3b.png)

as opposed to 

![image](https://user-images.githubusercontent.com/1944663/90889803-2299f400-e3fc-11ea-8454-5af6a3b3acaf.png)

astirpe replied:
It's ok. But I think the `to` should not be displayed is `date_time_end` is false

davejames replied:
I agree, I added that to the latest commit

astirpe reviewed (approved):
LGTM

davejames commented:
thanks @astirpe for the thorough review :)

davejames committed:
project_timesheet_time_control: add a field for the end time of a timesheet

This commit adds a field to the timesheet to store the end time. It is not required to enter the end time, if it is not entered it is calculated automatically by comparing the start time and the duration, if only a start time and end time are entered, then the duration is calculated accordingly. The end time field is set to optional in the tree view.

davejames commented:
squashed

pedrobaeza commented:
Thanks, merging!

/ocabot merge minor

Can you make a PR for changing the README of `hr_timesheet_activity_begin_end` for deprecating it saying that it's not needed using this one?
